profiler/src/ProfilerEngine/Datadog.Profiler.Native/FrameStore.cpp

```diff
@@ -896,6 +928,182 @@ std::pair<std::string, mdTypeDef> FrameStore::GetMethodNameFromMetadata(IMetaDat
     return std::make_pair(shared::ToString(shared::WSTRING(buffer.get())), mdTokenType);
 }
 
+std::string FrameStore::GetMethodSignature(ICorProfilerInfo4* pInfo, IMetaDataImport2* pMetaData, FunctionID functionId, mdMethodDef mdTokenFunc)
+{
+    PCCOR_SIGNATURE pSigBlob;
+    ULONG blobSize, attributes;
+    DWORD flags;
+    ULONG codeRva;
+
+    // get the coded signature from metadata
+    ULONG nameCharCount = 0;
+    HRESULT hr = pMetaData->GetMethodProps(mdTokenFunc, nullptr, nullptr, 0, nullptr, &attributes, &pSigBlob, &blobSize, &codeRva, &flags);
+    if (FAILED(hr))
+    {
+        return "(?)";
+    }
+
+    // read https://chnasarre.medium.com/decyphering-method-signature-with-clr-profiling-api-8328a72a216e for more details
+    ULONG elementType;
+    ULONG callConv;
+    std::stringstream builder;
+
+    // get the calling convention
+    pSigBlob += CorSigUncompressData(pSigBlob, &callConv);
+
+    ULONG argCount = 0;
+    ClassID classId = 0;
+    ModuleID moduleId;
+    // for generic support
+    std::unique_ptr<ClassID[]> methodTypeArgs = nullptr;
+    std::unique_ptr<ClassID[]> classTypeArgs = nullptr;
+    ULONG genericArgCount = 0;
+    UINT32 methodTypeArgCount = 0;
+    ULONG32 classTypeArgCount = 0;
+    if ((callConv & IMAGE_CEE_CS_CALLCONV_GENERIC) != 0)
+    {
+        //
+        // Grab the generic type argument count
+        //
+        pSigBlob += CorSigUncompressData(pSigBlob, &genericArgCount);
+
+        // get the generic details for the function
+        methodTypeArgs = std::make_unique<ClassID[]>(genericArgCount);
+        hr = pInfo->GetFunctionInfo2(functionId, NULL, &classId, &moduleId, NULL, genericArgCount, &methodTypeArgCount, methodTypeArgs.get());
+        assert(!SUCCEEDED(hr) || (genericArgCount == methodTypeArgCount));
```

Comment:
maybe use `FAILED(hr)` instead of `!SUCCEEDED(hr)`

Reply:
✅ done